### PR TITLE
Async event sending

### PR DESCRIPTION
**Pull request by FriedSock**
## Why?

We want to be able to publish events to routemaster in background job so that we don't lose updates if routemaster goes down/ we don't lose availability

## What?

Allow `routemaster-client` to be configured with sidekiq to send events in background jobs. We should do this in `routemaster-client` because the code should be exactly the same across any service that wants to use it

### CODE REVIEW

File: routemaster/client.rb [hunk 1]
```diff
@@ -9,17 +11,21 @@
...
-    
+
     def initialize(options = {})
-      @_url = _assert_valid_url(options[:url])
-      @_uuid = options[:uuid]
-      @_timeout = options.fetch(:timeout, 1)
-      @_verify_ssl = options.fetch(:verify_ssl, true)
+      @_options = options.tap do |o|
+        o[:timeout] ||= 1
+        o[:worker_type] ||= :null
```
Comment (nronas) [inline routemaster/client.rb]: why we cast a custom `:null`? and not just `nil`

File: routemaster/client.rb [hunk 2]
```diff
@@ -9,17 +11,21 @@
...
 module Routemaster
   class Client
-    
+
     def initialize(options = {})
-      @_url = _assert_valid_url(options[:url])
-      @_uuid = options[:uuid]
-      @_timeout = options.fetch(:timeout, 1)
-      @_verify_ssl = options.fetch(:verify_ssl, true)
+      @_options = options.tap do |o|
```
Comment (nronas) [inline routemaster/client.rb:16]: As far as i can see the `@_options` has local scope in this method so probably we can use just a local var here.

File: routemaster/client/connection.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        h.keys.each do |k|
+          h[k.to_s] = h.delete(k)
+        end
+      end
+    end
+
+    def _conn
+      @_conn ||= Faraday.new(@_url, ssl: { verify: @_verify_ssl }) do |f|
+        f.request :retry, max: 2, interval: 100e-3, backoff_factor: 2
+        f.request :basic_auth, @_uuid, 'x'
```
Comment (nronas) [inline routemaster/client/connection.rb]: what's the purpose of `x` here?

Comment (FriedSock) [reply routemaster/client/connection.rb]: no idea, copypasta method from the old code

File: routemaster/client.rb (hunk 2 above)
Comment (FriedSock) [reply routemaster/client.rb:16]: It is passed to the connection later on, so we need to save it.

File: routemaster/client.rb [hunk 4]
```diff
@@ -9,17 +11,22 @@
...
+        o[:timeout] ||= 1
+        o[:backend_type] ||= Backends::Synchronous
+      end
+
+      @_backend_type = @_options.fetch(:backend_type)
```
Comment (angetheo) [inline routemaster/client.rb:21]: You can avoid tapping before and just pass the default backend if they key is missing:
```
@_backend_type = @_options.fetch(:backend_type, Backends::Syncronous)
```
same goes for `@_timeout` which is actually in the previous implementation.

**Review by angetheo — APPROVED** [review]
nitpicks, else it LGTM

**Comment by nronas** [comment]
LGTM

File: routemaster/client.rb (hunk 4 above)
Comment (FriedSock) [reply routemaster/client.rb:21]: Yeah, kinda annoying that the timeout needs to be passed to the connection as part of the options hash though, seemed to make sense to do all the defaulting in the same way.

File: routemaster/client/backends.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+module Routemaster
+  class Client
+    module Backends
+      NAMES = ["Routemaster::Client::Backends::Synchronous", "Routemaster::Client::Backends::Sidekiq"]
+
+      autoload 'Synchronous', 'routemaster/client/backends/synchronous'
```
Comment (pedrocunha) [inline routemaster/client/backends.rb:6]: Are you guys aware of this? https://bugs.ruby-lang.org/issues/5653

Comment (FriedSock) [reply routemaster/client/backends.rb:6]: I wasn't, thanks, ill look to remove it in the next version.

**Pull request merged**